edx_oauth_client/backends/edx_oauth_client.py

```diff
@@ -21,117 +30,153 @@
 ]
 
 
-class EdxOAuthBackend(BaseOAuth2):
+class GenericOAuthBackend(BaseOAuth2):
     """
-    Backend for Edx OAuth Server Authorization
+    Backend for Edx OAuth Server Authorization.
     """
-    OAUTH_PROVIDER_URL = settings.FEATURES.get('OAUTH_PROVIDER_URL')
     name = 'edx-oauth2'
-    ID_KEY = settings.FEATURES.get('OAUTH_ID_KEY', '_id')
-    AUTHORIZATION_URL = '{}/oauth2/authorize'.format(OAUTH_PROVIDER_URL)
-    ACCESS_TOKEN_URL = '{}/oauth2/token'.format(OAUTH_PROVIDER_URL)
-    DEFAULT_SCOPE = settings.FEATURES.get('OAUTH_SCOPE', [])
-    REDIRECT_STATE = False
-    ACCESS_TOKEN_METHOD =  settings.FEATURES.get('OAUTH_ACCESS_TOKEN_METHOD', 'POST')
+    skip_email_verification = True
 
     PIPELINE = DEFAULT_AUTH_PIPELINE
+    REDIRECT_STATE = False
 
-    skip_email_verification = True
+    def authorization_url(self):
+        return self.setting("AUTHORIZATION_URL")
+
+    def access_token_url(self):
+        return self.setting("ACCESS_TOKEN_URL")
 
-    def setting(self, name, default=None):
-        """Return setting value from strategy"""
+    def setting(self, name, default=None, backend=None):
+        """
+        Load the setting from a ConfigurationModel if possible, or fall back to the normal
+        Django settings lookup.
+
+        OAuthAuth subclasses will call this method for every setting they want to look up.
+        """
+
+        provider_config = third_party_auth.models.OAuth2ProviderConfig.objects.filter(
+            backend_name=self.name,
+            site=Site.objects.get_current(get_current_request()),
+            enabled=True
+        ).last()
+
+        if provider_config and not provider_config.enabled_for_current_site:
+            raise Exception("Can't fetch setting of a disabled backend/provider.")
         try:
-            import third_party_auth
-        except ImportError:
-            OAuth2ProviderConfig = None
-
-        if third_party_auth.models.OAuth2ProviderConfig is not None:
-            providers = [p for p in third_party_auth.provider.Registry.displayed_for_login() if p.backend_name == self.name]
-            if not providers:
-                raise Exception("Can't fetch setting of a disabled backend.")
-            provider_config = providers[0]
-            try:
-                return provider_config.get_setting(name)
-            except KeyError:
-                pass
-        return super(EdxOAuthBackend, self).setting(name, default=default)
+            return provider_config.get_setting(name)
+        except KeyError:
+            pass
+
+        # Special case handling of login error URL if we're using a custom auth entry point:
```

Comment:
A comment is the last option. This is always bad and should be avoided in your code.
What we have here. We have a `setting` function which is super big and does a lot. This function has a comment where a special case is handled. Such a situation is a sign and a marker that this case needs to be moved to another function with a good name, and your comment in the code, if needed, would be an excellent docstring.

Reply:
Here It's a parent method implementation. But for generic realization we can not call super of the parent class because it will rise the exception. So that's why we need to process it here.

Parent method link: https://github.com/raccoongang/edx-platform/blob/ficus-rg/common/djangoapps/third_party_auth/strategy.py#L36

Reply:
I have nothing against it. But what's stopping you from taking this special case into a separate method with a good name and calling it here. But as you wish.